File: God Zuo/class057 并查集下/My03.py
# 测试链接 : https://leetcode.cn/problems/number-of-good-paths/

class Solution:
    def numberOfGoodPaths(self, vals: list[int], edges: list[list[int]]) -> int:
        n = len(vals)
        father=[_ for _ in range(n)]
        maxcnt=[1]*n

        def find(i):
            while  i != father[i]:
                father[i]=father[father[i]]
                i=father[i]
            return i
        
        ans=n
        edges.sort(key=lambda x : max(vals[x[0]],vals[x[1]]))

        for u,v in edges:
            fx=find(u)
            fy=find(v)
            if fx != fy:
                if vals[fx]>vals[fy]:
                    father[fy]=father[fx]
                elif vals[fx]<vals[fy]:
                    father[fx]=father[fy]
                else:
                    ans+=maxcnt[fx]*maxcnt[fy]
                    maxcnt[fx]+=maxcnt[fy]
                    father[fy]=fx
        return ans

File: God Zuo/class069 三维动态规划/My03.cpp
// 测试链接 : https://leetcode.cn/problems/knight-probability-in-chessboard/
class Solution{
public:
    double dp[30][30][110];
    int dx[8] = {-2, -1, 1, 2, 2, 1, -1, -2};
    int dy[8] = {1, 2, 2, 1, -1, -2, -2, -1};
    double f(int n,int i,int j,int k){
        if(i<0 || i>=n || j<0 || j>=n)
            return 0;
        if(dp[i][j][k]!=-1.0)
            return dp[i][j][k];
        double ans = 0;
        if(k==0){
            ans = 1;
        }else{
            for (int t = 0; t < 8;++t){
                ans += (f(n, i + dx[t], j + dy[t], k - 1) /8.0);
            }
        }
        dp[i][j][k] = ans;
        return ans;
    }
    double knightProbability(int n, int k, int row, int column){
        for (int i = 0; i < 30;++i){
            for (int j = 0; j < 30;++j){
                for (int l = 0; l < 100;++l)
                    dp[i][j][l] = -1;
            }
        }
        return f(n, row, column, k);
    }
};